src/radical/pilot/agent/executing/base.py

```diff
@@ -117,7 +125,54 @@ def work(self, tasks):
     #
     def command_cb(self, topic, msg):
 
-        raise NotImplementedError('work is not implemented')
+        raise NotImplementedError('command_cb is not implemented')
+
+
+    # --------------------------------------------------------------------------
+    #
+    def cancel_task(self, task):
+
+        raise NotImplementedError('cancel_task is not implemented')
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _to_watcher(self):
+        '''
+        watch the set of tasks for which timeouts are defined.  If the timeout
+        passes and the tasks are still active, kill the task via
+        `self._cancel_task(task)`.  That has to be implemented by al executors.
+        '''
+
+        while True:
+
+            # check once per second at most
+            time.sleep(1)
+
+            now = time.time()
+            with self._to_lock:
+
+                # running tasks for next check
+                to_list = list()
+                for to, start, task in self._to_tasks:
+                    if now - start > to:
+                        self._prof.prof('task_timeout', uid=task['uid'])
+                        self.cancel_task(task)
```

Comment:
for this part, should we have `_cancel_lock`? I mean to have it within base class, but Popen also will use it whenever method `cancel_task` is called?

Reply:
Hey @mtitov - I refactored the code quite a bit: cancel requests are now pushed into the watch queue, negating the need for a thread-shared variable and locks.  Let me know if that works for you.

On the down side that makes the tests a bit tricky AFAICT - do you have a suggestion on how the failing test could be restructured?  Should we just check if the next `get` on the watch queue yields the task ID?